presto-native-execution/presto_cpp/main/common/prometheus-metrics/PrometheusReporter.cpp

```diff
@@ -0,0 +1,172 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "PrometheusReporter.h"
+namespace facebook::presto::prometheus {
+void PrometheusReporter::registerMetricExportType(
+    const char* key,
+    facebook::velox::StatType statType) const {
+  if (registeredMetrics_.count(key)) {
+    // Already registered;
+    VLOG(1) << "Trying to register already registered metric " << key;
+    return;
+  }
+  // Prometheus format requires '.' to be replaced.
+  std::string keyStr = std::string(key);
+  std::replace(keyStr.begin(), keyStr.end(), '.', '_');
+  switch (statType) {
+    case facebook::velox::StatType::COUNT: {
+      auto& counterFamily =
+          ::prometheus::BuildCounter().Name(keyStr).Register(*registry_);
+      auto& counter = counterFamily.Add(labels_);
+      countersMap_.emplace(std::string(key), counter);
+      break;
+    }
+    case facebook::velox::StatType::SUM:
+    case facebook::velox::StatType::AVG:
+    case facebook::velox::StatType::RATE: {
+      auto& gaugeFamily =
+          ::prometheus::BuildGauge().Name(keyStr).Register(*registry_);
+      auto& gauge = gaugeFamily.Add(labels_);
+      gaugeMap_.emplace(std::string(key), gauge);
+      break;
+    }
+    default:
+      VELOX_UNSUPPORTED(
+          "Unsupported metric type {}", std::to_string((int)statType));
+  }
+  registeredMetrics_.emplace(key, statType);
+}
+
+void PrometheusReporter::registerMetricExportType(
+    folly::StringPiece key,
+    facebook::velox::StatType statType) const {
+  registerMetricExportType(key.start(), statType);
+}
+
+void PrometheusReporter::registerHistogramMetricExportType(
+    const char* key,
+    int64_t bucketWidth,
+    int64_t min,
+    int64_t max,
+    const std::vector<int32_t>& pcts) const {
+  if (registeredHistogramMetrics_.count(key)) {
+    // Already registered;
+    VLOG(1) << "Trying to register already registered metric " << key;
+    return;
+  }
+  int numBuckets = (max - min) / bucketWidth;
+  auto bound = min + bucketWidth;
+  std::string keyStr = std::string(key);
+  std::replace(keyStr.begin(), keyStr.end(), '.', '_');
+  auto& histogramFamily =
+      ::prometheus::BuildHistogram().Name(keyStr).Register(*registry_);
+  ::prometheus::Histogram::BucketBoundaries bucketBoundaries;
+  while (numBuckets) {
+    bucketBoundaries.push_back(bound);
+    bound += bucketWidth;
+    numBuckets--;
+  }
+  VELOX_CHECK_GE(bucketBoundaries.size(), 1);
+  auto& histogramMetric = histogramFamily.Add(labels_, bucketBoundaries);
+  histogramMap_.emplace(key, histogramMetric);
+  // If percentiles are provided, create a Summary type metric and register.
+  if (pcts.size() > 0) {
+    auto& summaryFamily = ::prometheus::BuildSummary()
+                              .Name(keyStr.append(kSummarySuffix))
```

Comment:
Wont the metric name changed if we add a ksummarySuffix? Why is this needed?

Reply:
Or is it so that maybe we don't need the histogram if pcts are present and we can just register summary?

Reply:
If you check prometheus, Histograms and Summaries are 2 different metric types. If the incoming metric has quantiles (pcts) specified, then we use summary type as well.
https://prometheus.io/docs/concepts/metric_types/#histogram

Reply:
https://prometheus.io/docs/practices/histograms/#:~:text=The%20essential%20difference%20between%20summaries,the%20server%20side%20using%20the
`The essential difference between summaries and histograms is that summaries calculate streaming φ-quantiles on the client side and expose them directly, while histograms expose bucketed observation counts and the calculation of quantiles from the buckets of a histogram happens on the server side using the [histogram_quantile() function](https://prometheus.io/docs/prometheus/latest/querying/functions/#histogram_quantile).`
So maybe we do not need to register histogram if quantiles are present?

Reply:
Not sure about this^ just fyi

Reply:
We can revisit this. Not sure at this point if both give different viewpoints or Summary is sufficient over histograms.

Reply:
Cool thats fine, just check the `kSummarySuffix` i think the code here should be keyStr + "_Suffix" by value and we should not change the reference here